docs/BuildCaching.md

```diff
@@ -0,0 +1,30 @@
+Build Caching
+=============
+
+This repository can build the entire ASP.NET Core stack from scratch. However, in most cases, we only want to build part of the stack.
+
+## Building with cached artifacts
+
+This is the default behavior when `build.ps1` without arguments.
+
+This repository may contains files under `releases/*.xml`. These files follow the [Bill of Materials](./BillOfMaterials.md) format.
+Any artifacts described in these bom's are assumed to have been built in a previous release and are available for download.
+The build may use this information to skip building certain assets when a build action would produce a duplicate asset.
+
+## Disabling caching
+
+You can for the build to skip reading the bill of materials and build everything from scratch by calling `build.ps1 -NoCache`.
+This will cause the build to always rebuild assets.
+
+## When to commit a new bill of material
+
+Each build will produce a new Bill of Materials (bom). Normally, this bom isn't added to source code right away.
+This bom should only be committed to source once the assets it describes are published and available from caching mechanisms.
+
+## Caching mechanisms
+
+These are some common caching mechanisms used to store build artifacts.
```

Comment:
I think there needs to be a way in the bom to specify where each artifact is cached. Also what if there are multiple possibilities? The installers are generally uploaded to the azure blob and then released on debian and rpm feeds later on. For example, the store archives are also available through the download center.

Reply:
Yes. This kind of configuration should exist in the repo, but not in the bill of materials. We do something like this today with nuget packages. I could easily imagine expanding this approach to include blob feeds too. What this *might* look like:

```xml
   <PropertyGroup>

    <RestoreSources>$(DotNetRestoreSources)</RestoreSources>
    <RestoreSources Condition="'$(DotNetBuildOffline)' != 'true'  ">
      $(RestoreSources);
      https://dotnet.myget.org/F/aspnetcore-dev/api/v3/index.json;
      https://api.nuget.org/v3/index.json;
    </RestoreSources>

    <BlobSources>$(DotNetBlobSources)</BlobSources>
    <BlobSources Condition="'$(DotNetBuildOffline)' != 'true'  ">
      $(BlobSources);
      https://aspnetcore.blob.core.windows.net/buildfeed/;
      https://dist.asp.net/buildfeed/
      file:///$(MSBuildThisFileDirectory)/../.deps/;
    </AzureBlobSources>

  </PropertyGroup>
```